hail/src/main/scala/is/hail/types/physical/PCanonicalBaseStruct.scala

```diff
@@ -224,54 +166,45 @@ abstract class PCanonicalBaseStruct(val types: Array[PType]) extends PBaseStruct
     }
   }
 
-  override def load(src: Code[Long]): PCanonicalBaseStructCode =
-    new PCanonicalBaseStructCode(this, src)
-}
-
-object PCanonicalBaseStructSettable {
-  def apply(sb: SettableBuilder, pt: PCanonicalBaseStruct, name: String): PCanonicalBaseStructSettable = {
-    new PCanonicalBaseStructSettable(pt, sb.newSettable(name))
-  }
-}
-
-class PCanonicalBaseStructSettable(
-  val pt: PCanonicalBaseStruct,
-  val a: Settable[Long]
-) extends PBaseStructValue with PSettable {
-  def get: PBaseStructCode = new PCanonicalBaseStructCode(pt, a)
-
-  def settableTuple(): IndexedSeq[Settable[_]] = FastIndexedSeq(a)
+  def sType: SStruct = SBaseStructPointer(this)
 
-  def loadField(cb: EmitCodeBuilder, fieldIdx: Int): IEmitCode = {
-    IEmitCode(cb,
-      pt.isFieldMissing(a, fieldIdx),
-      pt.fields(fieldIdx).typ.load(pt.fieldOffset(a, fieldIdx)))
-  }
-
-  def store(pv: PCode): Code[Unit] = {
-    a := pv.asInstanceOf[PCanonicalBaseStructCode].a
-  }
+  def getPointerTo(cb: EmitCodeBuilder, addr: Code[Long]): PCode = SBaseStructPointer(this).loadFrom(cb, null, this, addr)
 
-  def isFieldMissing(fieldIdx: Int): Code[Boolean] = {
-    this.pt.isFieldMissing(a, fieldIdx)
+  def store(cb: EmitCodeBuilder, region: Value[Region], value: PCode, deepCopy: Boolean): Code[Long] = {
+    value.st match {
+      case SBaseStructPointer(t) if t.equalModuloRequired(this) && !deepCopy =>
+        value.asInstanceOf[SBaseStructPointerCode].a
+      case _ =>
+        val newAddr = cb.newLocal[Long]("pcbasestruct_store_newaddr")
+        cb.assign(newAddr, allocate(region))
+        storeAtAddress(cb, newAddr, region, value, deepCopy)
+        newAddr
+    }
   }
-}
-
-class PCanonicalBaseStructCode(val pt: PCanonicalBaseStruct, val a: Code[Long]) extends PBaseStructCode {
-  def code: Code[_] = a
 
-  def codeTuple(): IndexedSeq[Code[_]] = FastIndexedSeq(a)
-
-  def memoize(cb: EmitCodeBuilder, name: String, sb: SettableBuilder): PBaseStructValue = {
-    val s = PCanonicalBaseStructSettable(sb, pt, name)
-    cb.assign(s, this)
-    s
+  def storeAtAddress(cb: EmitCodeBuilder, addr: Code[Long], region: Value[Region], value: PCode, deepCopy: Boolean): Unit = {
+    value.st match {
+      case SBaseStructPointer(t) if t.equalModuloRequired(this) =>
+        val pcs = value.asBaseStruct.memoize(cb, "pcbasestruct_store_src").asInstanceOf[SBaseStructPointerSettable]
+        val addrVar = cb.newLocal[Long]("addr", addr)
+        cb += Region.copyFrom(pcs.a, addrVar, byteSize)
+        if (deepCopy)
+          deepPointerCopy(cb, region, addrVar)
+      case _ =>
+        val addrVar = cb.newLocal[Long]("addr", addr)
```

Comment:
`"addr"` is not a sufficiently unique name, let's use something more searchable / inline with `pcbasestruct_store_src`.

Reply:
👍